### PR TITLE
Scroll Using Click-and-Drag with Middle Mouse Button

### LINKED ISSUE #1505: Beginn scrolling by clicking middle mouse button
**Is your feature request related to a problem? Please describe.**

A lot of software (emg. Browsers) start scrolling when clicking the middle mouse button. Chatterino would be more in line when it also supported that.

**Additional context**

Behaviour should be similar to the implementation in browsers.

### PR DESCRIPTION
**Pull request by leon-richardt**
### Description
Allows a behavior similar to web browser behavior: After clicking the middle mouse button, a user can move the mouse up or down to scroll the chat in the desired direction.
Resolves #1505.

### Additional Questions & Considerations
- [x] Make scrolling behave consistently with different screen scales.
- [x] Figure out good trade-off between `scrollTimer_` interval and UX.
- [x] Use a custom cursor pixmap for showing different scroll states, like used in browsers.

### CODE REVIEW

File: src/widgets/helper/ChannelView.cpp [hunk 1]
```diff
@@ -130,6 +130,27 @@ ChannelView::ChannelView(BaseWidget *parent)
     this->clickTimer_->setSingleShot(true);
     this->clickTimer_->setInterval(500);
 
+    // TODO(leon): What interval to choose?
+    this->scrollTimer_.setInterval(20);
+    QObject::connect(&this->scrollTimer_, &QTimer::timeout, this, [this] {
+        const qreal delta = this->currentMousePosition_.y() -
+                            this->lastMiddlePressPosition_.y();
+
+        // TODO(leon): How large should the "grace area" be?
```
Comment (fourtf) [inline src/widgets/helper/ChannelView.cpp]: doesn't account for scale, so if you're using 4k then it's only 5 px

**Comment by fourtf** [comment]
It doesn't visualize the starting point like in chrome which not expected and bad for learnability (user doesn't know what's going on)

**Comment by leon-richardt** [comment]
> Use a custom cursor pixmap for showing different scroll states, like used in browsers.

That's what this TODO was supposed to describe, guess I didn't word it very nicely. It's definitely ooj the list though. 👍

**Comment by fourtf** [comment]
20 for the interval seems fine to me

File: src/widgets/helper/ChannelView.cpp (hunk 1 above)
Comment (leon-richardt) [reply src/widgets/helper/ChannelView.cpp]: I believe that this is fixed by 8f2db4d. Can't test though because I don't have a high DPI monitor.

**Comment by leon-richardt** [comment]
> Chat line height affects the scrolling speed.
> If all chat lines are short enough to fit on one line in your window, the scrolling will be smooth.
> If some chat lines span more than one line, the scrolling over that line will be jagged.

Any idea on what might be causing this?

**Comment by pajlada** [comment]
> > Chat line height affects the scrolling speed.
> > If all chat lines are short enough to fit on one line in your window, the scrolling will be smooth.
> > If some chat lines span more than one line, the scrolling over that line will be jagged.
> 
> Any idea on what might be causing this?

It looks to be reproducible when using the scrollbar like normal too, so I wouldn't say this is a blocker to merge the PR.

**Comment by leon-richardt** [comment]
Left and right clicks now also disable scrolling.

**Comment by cinemassacres** [comment]
the dev of ffz is talking to the dev of bttv  a bit about getting overlay emotes added to the BTTV API. so we won't need this hack  If that happens, then overlay emotes will start to just work.

**Comment by leon-richardt** [comment]
> the dev of ffz is talking to the dev of bttv a bit about getting overlay emotes added to the BTTV API. so we won't need this hack If that happens, then overlay emotes will start to just work.

I believe you meant to post this on #1613?

**Comment by pajlada** [comment]
Works great for me, the stutter on different message heights is still a thing but it's a scrollbar thing, which I think isn't up to this PR to fix

**Pull request merged**